### ISSUE TITLE
`Lrnr_rpart` fails to predict on external datasets

### ISSUE THREAD

**Issue by WenxinZhang25** [tlverse/sl3#319]
Hello,

I met a problem that I cannot predict an external dataset (5 covariates, n = 300 or 1000) with the Lrnr_rpart model trained on a dataset with 1000 rows. 

I found two main problems:
1. Every covariate's name begins with an additional "x" in the tree of `rpart_fit` . I think there might be some bugs in the variable name in Lrnr_rpart which triggers some problems for stat::predict to match the format of input data. In this case, stat::predict only returns 1000 predictions on training set, not the given external test set.  (See https://stackoverflow.com/questions/27464893/getting-warning-newdata-had-1-row-but-variables-found-have-32-rows-on-pred/27465510)

2. If we look at rpart_fit$fit_object$terms, the dataClasses of x is a matrix. It might imply that our newdata must be a matrix with the same dimension, which does not make sense either.
```
lrnr_rpart = Lrnr_rpart$new()
rpart_fit = lrnr_rpart$train(task)
length(predict(rpart_fit$fit_object,task_test_300$X))
```

> 'newdata' had 300 rows but variables found have 1000 rows[1] 1000

```
rpart_fit$fit_object$terms
```

>...
attr(,"dataClasses")
          y           x 
  "numeric" "nmatrix.5"

```
rpart_fit
```
> [1] "Lrnr_rpart_FALSE_FALSE_FALSE"

> n= 1000 
> 
> node), split, n, deviance, yval
>       * denotes terminal node
> 
>  1) root 1000 931.683900 -0.0007554607  
>    2) xx5>=-0.1654207 586 527.846000 -0.0756749800 *
>    3) xx5< -0.1654207 414 395.893000  0.1052900000  
>      6) xx5< -0.994509 151 130.125900 -0.1775084000  
>       12) xx1< 1.122358 132 111.141300 -0.2723032000 *
>       13) xx1>=1.122358 19   9.557768  0.4810660000 *
>      7) xx5>=-0.994509 263 246.757400  0.2676572000 *

If I directly use rpart::rpart method on a matrix of X (as it is in _**Line 71 of sl3/R/Lrnr_rpart.R**_), everything looks good, without problems mentioned before. 

```{r}
library(rpart)
## imitating the code in line 71 in sl3/R/Lrnr_rpart.R
x <- as.matrix(task$X)
y <- task$Y
formula <- data.frame(y = y, x)
fit = rpart(formula = formula)
# No problem in prediction
length(stats::predict(fit, newdata = test_data_300))

```
> [1] 300

```{r}
fit$terms
 ```

> ...
> attr(,"dataClasses")
>         y        x1        x2        x3        x4        x5 
> "numeric" "numeric" "numeric" "numeric" "numeric" "numeric" 

```{r}
fit
 ```

> n= 1000 
> 
> node), split, n, deviance, yval
>       * denotes terminal node
> 
>  1) root 1000 931.683900 -0.0007554607  
>    2) x5>=-0.1654207 586 527.846000 -0.0756749800 *
>    3) x5< -0.1654207 414 395.893000  0.1052900000  
>      6) x5< -0.994509 151 130.125900 -0.1775084000  
>       12) x1< 1.122358 132 111.141300 -0.2723032000 *
>       13) x1>=1.122358 19   9.557768  0.4810660000 *
>      7) x5>=-0.994509 263 246.757400  0.2676572000 *

Additionally, even if my input matrix is exactly 1000 * 5, the predictions are still for the training data.

```{r}

length(predict(rpart_fit$fit_object, 
               newdata = data.frame(y = task_test_1000$Y, 
                                    x = as.matrix(task_test_1000$X))))

all(predict(rpart_fit$fit_object, 
            newdata = data.frame(y = task_test_1000$Y, 
                                 x = as.matrix(task_test_1000$X))) 
    == rpart_fit$predict())

all(rpart_fit$predict(task_test_1000) 
    == rpart_fit$predict())


```

> [1] 1000

> [1] TRUE
> [1] TRUE

**Comment by jeremyrcoyle** [comment]
Hi Wenxin,

I can't replicate either issue you're reporting. Below is an example confirming that you can generate external predictions and that the term labels are X1-X5 (which is how they are in the training data). Please confirm that it works for you as well.

```
context("test-Lrnr_rpart.R -- General testing for Rpart")

library(sl3)
library(testthat)
library(rpart)

gendata <- function(n=1000,p=5){
  x <- matrix(rnorm(n*p),nrow=n)
  ey <- x[,1] + x[,2] 
  y <- ey + rnorm(n,0,0.1)
  
  data.frame(x,ey, y)
}

data <- gendata()
covariates <- setdiff(names(data),c("y","ey"))
task <- make_sl3_Task(data, covariates = covariates, outcome="y")
lrnr_rpart <- make_learner(Lrnr_rpart)
fit <- lrnr_rpart$train(task)
preds <- fit$predict()

# check fit tree names
fit$fit_object$terms

# predict on external data
external <- gendata(300)
ext_task <- make_sl3_Task(external, covariates = covariates, outcome="y")
ext_preds <- fit$predict(ext_task)
mean((external$ey-ext_preds)^2)
```

**Comment by WenxinZhang25** [comment]
Hi Jeremy,

I still had those problems before I upgraded my sl3 from 1.3.7 to 1.4.2 just now. Now everything works well! Thank you very much!

**Comment by nhejazi** [comment]
Great, looks like this is resolved --- closing.
